addons/web/controllers/export.py

```diff
@@ -161,8 +159,8 @@ def insert_leaf(self, group):
             # Update count value and aggregated value.
             node.count += count
 
-        node.data = records.with_env(self._model.env).export_data(self._export_field_names).get('datas', [])
-        return records
+        record_ids = group.pop('id:array_agg')
+        node.data = [row for record_id in record_ids for row in self._export_data[record_id]]
```

Comment:
With your current code, you take the order of `id:array_agg` (order by id, not the natural order of the model). However the `_export_data` is in the right order.

I bet that all the test is working fine since the _order of models used is always `id` (which is by chance the same logic done in `id:array_agg`.

```suggestion
        record_ids = set(group.pop('id:array_agg'))
        node.data = [row for record_id, row in self._export_data.items() if record_id in record_ids]
```
But the worst case become O = (r²) (r=record, if you have one group by record), it may be worth it to change `_export_data` to only contains `id:array_agg` records. I don't know if my comment makes sense.